---
title: 'Proteomics vignette: dimensionality reduction and imputation 
with sgdGMF'
author:
- name: Alexandre Segers
bibliography: sgdGMF.bib
date: "14/01/2025"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
  BiocStyle::pdf_document: default
package: omicsGMF
abstract: |
  Proteomics vignette for the omicsGMF package. This vignette aims to provide a 
  detailed description of a matrix factorization done for proteomics data, 
  which can further be used for visualization and imputation of missing values.
vignette: >
  %\VignetteIndexEntry{Proteomics-vignette: omicsGMF}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r, echo = FALSE}
library(knitr)
```

# Introduction

`omicsGMF` is a R package to perform generalized matrix factorization which
can be used for dimensionality reduction and visualization of omics data, and 
even for the imputation of missing values. It can deal with count data and 
missing values in a fast way, and therefore 
does not require log-transformation of RNA-seq data or imputation of proteomics
data such as conventional PCA. It can perform model selection to optimize the 
number of latent confounders, i.e. the dimensionality of the reduced space. 
`omicsGMF` uses the `sgdGMF` package, but 
provides easy wrappers for `SingleCellExperiment` and `SummarizedExperiment` 
classes, with adapted default values of the parameters for omics data.

All details about the `sgdGMF` model, such as the adaptive learning rates,
exponential gradient averaging and subsampling of the data are 
described in our preprint [@Castliglioni2024]. There, we show the use of 
`sgdGMF` on single-cell RNA-seq data. In our other preprint [@Segers2025], 
we show how this framework can be used to visualize and impute missing values 
in (single-cell) proteomics data.

In this vignette, we show how `omicsGMF` can be easily used to obtain the
dimensionality reduction of omics data, to visualize these samples and how to 
impute missing values. It serves as an alternative to PCA when dealing with
count or incomplete data, while remaining fast due to its stochastic 
optimization algorithms. In total, there are four main functions:

1. `calculateCVGMF` performs cross-validation to determine the optimal number
of latent confounders. These results can be visualized using `plotCV`. This avoids arbitrarily choosing `ncomponents`, but
requires some computational time. An alternative is `calculateRankGMF`, which
performs an eigenvalue decomposition on the deviance residuals. This allows for
model selection based on a scree plot using `plotRank`, for example using
the elbow method.

2. `runGMF` or `calculateGMF` estimates the latent confounders and the rotation
matrix, and estimates the respective parameters of the sample-level and 
feature-level covariates.

3. `plotGMF` plots the samples using its decomposition.

4. `imputeGMF` creates a new assay with missing values imputes using the 
estimates of `runGMF`.


We first show a standard workflow of `omicsGMF` on RNA-seq data, followed by 
a standard workflow on proteomics data.

# Package installation

`sgdGMF` can be installed through CRAN.
`omicsGMF` can be installed from Bioconductor with:

```{r, eval=FALSE}
if(!requireNamespace("sgdGMF", quietly = TRUE))
    install.packages("sgdGMF")

if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("omicsGMF")
```


```{r, echo = TRUE, warning=FALSE, message=FALSE}
library(sgdGMF)
library(omicsGMF)
library(dplyr)
library(scuttle)
set.seed(100)
```


# Proteomics data analysis

For the proteomics data analysis, we will simulate some artificial Gaussian 
data. Also, we introduce some missing values completely ad random. Note that on 
real data, one might have to perform log-transformation prior to computing the
matrix factorization to make the intensities Gaussian distributed. Also, one 
can optionally opt to perform some normalization such as median-normalization.
We store the simulated intensities in the `logintensities` assay of a 
`SingleCellExperiment`.

```{r}
sim_intensities <- matrix(rnorm(n = 20*500, mean = 1, sd = 1),
                          ncol = 20)
NAs <- rbinom(n = prod(dim(sim_intensities)), size = 1, prob = 0.3) == 1
sim_intensities[NAs] <- NA

colnames(sim_intensities) <- paste0("S_", c(1:20))
rownames(sim_intensities) <-  paste0("G_", c(1:500))

example_sce <- SingleCellExperiment(assays = SimpleList("logintensities" = sim_intensities))
```


A recommended step is to estimate the optimal dimensionality in the model
by using cross-validation. This cross-validation masks a proportion of the
values as missing, and tries to reconstruct these. Using the out-of-sample 
deviances, one can estimate the optimal dimensionality of the latent space.
This can be done with the `calculateCVGMF` function, which uses 
\link{sgdgmf.cv} from the `sgdGMF` package. Although the `sgdGMF` package 
allows great flexibility regarding the optimization algorithm is possible, 
sensible default values are here introduced for omics data. One
should mainly choose the right distribution family (`family`) and the number
of components in the dimensionality reduction for which the cross-validation is
run (`ncomponents`). Also, one should select the right assay that is used for
dimensionality reduction (`exprs_values` or `assay.type`).


```{r}
cv_proteomics <- calculateCVGMF(example_sce, 
                                 exprs_values="logintensities", 
                                 family = gaussian(), 
                                 ncomponents = c(1:5))

cv_proteomics %>% group_by(ncomp) %>% summarise(mean_dev = mean(dev),
                                                mean_aic = mean(aic),
                                                mean_bic = mean(bic),
                                                mean_mae = mean(mae),
                                                mean_mse = mean(mse))
```


If data are very large, or when having no idea which range of cross-validation 
values to explore, one can also use a scree plot of the eigenvalues for model
selection. This approach uses PCA on the deviance residuals to obtain the
eigenvalues, which is an approximation to the real matrix factorization.
This can be done using`runRankGMF` or `calculateRankGMF` followed
by `plotRank` or `screeplot_rank` respectively. Note that now, the 
`maxcomp` parameter should be defined, which is the number of 
eigenvalues computed.

Note that missing values are here imputed with the feature means. 
Therefore, when dealing with many missing
values, using a scree plot for model selection might not be optimal, but 
however remains a fast alternative to cross-validation.

```{r}
rank_proteomics <- calculateRankGMF(example_sce, 
                      exprs_values="logintensities", 
                      family = gaussian(),
                      maxcomp = 10)

screeplot_rank(rank_proteomics, maxcomp = 10)
```


After choosing which number of components to use in the final dimensionality
reduction, `runGMF` or `calculateGMF` can be used. Again, one should select the
distribution family (`family`), the dimensionality (`ncomponents`) and the 
assay used (`exprs_values` or `assay.type`). Note that, differently compared to
`runPCA`, by default all features are used. If this is not wanted, one can 
for example choose the n most variable genes used with `ntop`. However,
when the goal is imputing missing values, all features should be used.
`runGMF` returns the dimensionality
reduction in the `reducedDim` argument of the `SingleCellExperiment`. 
Additional information such as the `rotation` matrix, parameter estimates, 
the optimization history of `sgdGMF` and many more are available in the 
`attributes`. See \link{runGMF} for all outputs. 

```{r}
example_sce <- runGMF(example_sce, 
                      exprs_values="logintensities", 
                      family = gaussian(), 
                      ncomponents = 3,
                      name = "GMF")
```


After computing the matrix factorization, one can visualize the results with
functions from the \link{scater} package. For example, using `plotReducedDim`,
one obtains a PCA plot, using the latent factors obtained by `omicsGMF`. The 
same plot can be obtained using the wrapper function `plotGMF`.


```{r}
plotReducedDim(example_sce, dimred = "GMF")
reducedDimNames(example_sce)
head(reducedDim(example_sce, type = "GMF"))
```

Also, it is possible to impute the missing values with the mean estimates. This
can be done with the `imputeGMF` function:

```{r}
example_sce <- imputeGMF(example_sce, 
                         exprs_values = "logintensities", 
                         reducedDimName = "GMF",
                         name = "logintensities_imputed")


assay(example_sce,'logintensities')[1:5,1:5]
assay(example_sce,'logintensities_imputed')[1:5,1:5]
```


```{r}
sessionInfo()
```

